### ISSUE TITLE
Xdocker not connecting to Daemon

### ISSUE THREAD

**Issue by rorygilfillan** [jhipster/jhipster-ide#177]
"Test Connection" in "Docker Properties" gives Error connecting to docker!.
I am using version Version 17.12.0-ce-win47 (15139) of Docker.
I tried enabling "Expose daemon on tcp://localhost:2375" in the settings but still can't connect.

**Comment by colameo** [comment]
I thing you are mixing two different things. With "Test connection" in the docker preference page you can only check the connection to a docker daemon and has nothing to do with exposing applications ports of a docker container. 

Did you have a look to my [video on youtube](https://www.youtube.com/watch?v=LERTahPqVjo) where I also explain how to use Xdocker?

**Comment by colameo** [comment]
...try the followings:
- stop your local docker daemon (Docker tools)
- open a shell termine (do not know if you're using *nix or windows but should be the same) 
- type `docker info`
$ docker info
Cannot connect to the Docker daemon at **unix:///var/run/docker.sock**. Is the docker daemon running?
- open Xdocker preference page and set in the "Docker URL" text field the above address 
- start your docker daemon
- click on "Test connection" again
- now you should get the message "Connection established successful"

I've just tried with the my local docker installation (v18.03.0) and the latest Xdocker plugin on OSX 10.12.6 and it works.

Let me know if it worked. 

BTW, we have also a gitter chatroom => https://gitter.im/jhipster/jhipster-ide

**Comment by rorygilfillan** [comment]
I stopped the docker and ran docker info. Got the following:
error during connect: Get http://%2F%2F.%2Fpipe%2Fdocker_engine/v1.35/info: open //./pipe/docker_engine: The system cannot find the file specified. In the default daemon configuration on Windows, the docker client must be run elevated to connect. This error may also indicate that the docker daemon is not running.

Neither "http://%2F%2F.%2Fpipe%2Fdocker_engine/v1.35/" nor "//./pipe/docker_engine" are able to connect once I restarted docker. This is in windows.

**Comment by colameo** [comment]
Sorry, but I do not have a Windows machine to test that...

However, after exposing the daemon port I guess you've also started the docker daemon right?

![settings-general](https://user-images.githubusercontent.com/629760/37976182-35383132-31e1-11e8-99a9-5e54258eb720.jpg)

Maybe you should try to uninstall docker, delete from your home `.docker` folder and re-install the newest docker tools version.

After that expose from docker tool settings tab the daemon port, start it and check again in Xdocker preference page if you can connect.

Hope this will help, otherwise please report again

**Comment by rorygilfillan** [comment]
I updated my docker and 'tried enabling "Expose daemon on tcp://localhost:2375" in the settings but still can't connect.'

**Comment by colameo** [comment]
sorry, for my late answer, but I was quite busy in the last couple of weeks. 

Can you try this: `tcp://docker.for.win.localhost:2375`

**Comment by rorygilfillan** [comment]
Still can't connect

**Comment by colameo** [comment]
Which windows version are you using?

**Comment by colameo** [comment]
...refer also [here](https://github.com/docker/toolbox/issues/636)

**Comment by rorygilfillan** [comment]
I'm on windows 10. Docker works fine from the command line. I can also build a docker image with mvnw verify -Pprod dockerfile:build, after exposing the daemon.

**Comment by colameo** [comment]
ok, finally I found the problem... the (generated) default setting is wrong!

![2018-05-19 22_36_39-ws - myapp_src_model_model jdl - eclipse](https://user-images.githubusercontent.com/629760/40272896-14aabbfc-5bb6-11e8-953d-a657e437ae20.png)

change according to **your exposed daemon port address** (in my case tcp://localhost:2375):

![2018-05-19 22_37_21-ws - myapp_src_model_model jdl - eclipse](https://user-images.githubusercontent.com/629760/40272907-439d346c-5bb6-11e8-863a-269bba47c106.png)

After changing the address, I could connect to the local docker daemon and list docker containers:

![2018-05-19 22_49_37-ws - myapp_src_model_model jdl - eclipse](https://user-images.githubusercontent.com/629760/40272940-45fc0bba-5bb7-11e8-8f5f-41219d94abe1.png)

Please try again and report if that works for you.

I will fix this in the next release - thanks again to report this issue!

**Comment by rorygilfillan** [comment]
Still doesn't work
![image](https://user-images.githubusercontent.com/35095992/40296074-fbc07632-5cdb-11e8-8766-68e2cf0dd237.png)

**Comment by colameo** [comment]
- have you started docker before?
- after changing the URL preference, did you click the apply button?

BTW, Docker on Windows has evolved a lot. I've noticed that when I've upgraded my VirtualBox VM to Windows 10 Pro & Docker for Windows. The new implementation uses another virtualisation layer (Hyper-V) which brings (in my opinion) more troubles than benefits... 

However, I've implemented Xdocker in 2014/2015 and it's time to make an upgrade of the underlying technologies and APIs anyway. So, I suggest to either wait for the next JHipster IDE release (1.4.0) by end of June or switch to an *nix OS ;-)
